kissim/comparison/tree.py

```diff
@@ -0,0 +1,246 @@
+
+    Returns
+    -------
+    tree : scipy.cluster.hierarchy.ClusterNode
+        Cluster node.
+    mean_similarity : dict of int: float
+        Mean distance (value) for each node index (key).
+
+    References
+    ----------
+    [1] https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html
+    """
+
+    outputfile = Path(outputfile)
+
+    # Removing problematic entries if they exist
+    for entry in PROBLAMATIC_KINASES:
+        if entry in distance_matrix:
+            distance_matrix.drop(entry, axis=0, inplace=True)
+            distance_matrix.drop(entry, axis=1, inplace=True)
+
+    # Curate diagonal - set to 0
+    np.fill_diagonal(distance_matrix.values, 0)
+
+    # Hierarchical clustering (Ward by default)
+    # Alternatives: 'complete', 'weighted', 'average', 'centroid'
+    print(f"Clustering (method: {clustering_method}) and calculating branch similarities")
+    hclust = sch.linkage(ssd.squareform(distance_matrix.values), method=clustering_method)
+    print("Converting clustering to a Newick tree")
+    tree = sch.to_tree(hclust, False)
+
+    # Calculate and assign mean similarity for each of the branches
+    mean_similarity = {}
+    _get_mean_index(tree, distance_matrix, mean_similarity)
+
+    # Save tree to file
+    if outputfile:
+        _to_newick(tree, mean_similarity, distance_matrix, outputfile)
+        _to_kinase_annotation(distance_matrix, outputfile.parent / "kinase_annotations.csv")
+
+    return tree, mean_similarity
+
+
+def _to_newick(tree, mean_similarity, distance_matrix, outputfile):
+    """
+    Save Newick tree to file.
+
+    Parameters
+    ----------
+    tree : scipy.cluster.hierarchy.ClusterNode
+        Cluster node.
+    mean_similarity : dict of int: float
+        Mean distance (value) for each node index (key).
+    distance_matrix : pandas.DataFrame
+        Distance matrix on which clustering is based.
+    outputfile : str or pathlib.Path
+        Path to kinase tree file (TREE file) in Newick format.
+    """
+
+    outputfile = Path(outputfile)
+
+    print(f"Writing resulting tree to {outputfile}")
+    newick = ""
+    newick = _get_newick(tree, newick, tree.dist, list(distance_matrix), mean_similarity)
+    with open(outputfile, "w") as f:
+        f.write(newick)
+
+
+def _to_kinase_annotation(distance_matrix, outputfile):
+    """
+    Save kinase annotations to file used for FigTree.
+
+    Parameters
+    ----------
+    distance_matrix : pandas.DataFrame
+        Distance matrix on which clustering is based.
+    outputfile : str or pathlib.Path
+        Path to kinase annotation file (CSV file) in FigTree format.
+    """
+
+    outputfile = Path(outputfile)
+
+    # Get kinase names from matrix
+    kinase_names = distance_matrix.columns.to_list()
+
+    # Query KLIFS for kinase details
+    klifs_session = klifs.setup_remote()
+    kinases = klifs_session.kinases.by_kinase_name(kinase_names)
```

Comment:
An alternative is to speed this call up by calling this function without any kinase names and afterwards filter the returned dataset. 
This will save you a few hundred API calls on every run -> a lot speedier ;-)

Reply:
Thank a lot for your quick and thorough review! 

I am unsure why I implemented this kinase query in `opencadd` per-kinase, does not make sense to me anymore. 
I updated `opencadd.databases.klifs` accordingly - while doing that I might have stumbled over a tiny bug in the KLIFS API, please check the PR https://github.com/volkamerlab/opencadd/pull/75